# Introduction to ChIP-seq using high performance computing


| Audience | Computational Skills | Prerequisites | Duration |
:----------|:----------|:----------|:----------|
| Biologists | Beginner/Intermediate | None | 3-day workshop (~19.5 hours of trainer-led time)|

### Description

This repository has teaching materials for a 3-day Introduction to ChIP-sequencing data analysis workshop. This workshop focuses on teaching basic computational skills to enable the effective use of an high-performance computing environment to implement a ChIP-seq data analysis workflow. It includes an introduction to shell (bash) and shell scripting. In addition to running the ChIP-seq workflow from FASTQ files to peak calls and nearest gene annotations, the workshop covers best practice guidlelines for ChIP-seq experimental design and data organization/management and quality control.

> These materials were developed for a trainer-led workshop, but are also amenable to self-guided learning.

### Learning Objectives

1.	Understand the necessity for, and use of, the command line interface (bash) and HPC for analyzing high-throughput sequencing data.
2.	Understand best practices for designing a ChIP-seq experiment and analysis the resulting data.


### Contents

| Lessons            | Estimated Duration |
|:------------------------|:----------|
|[Introduction to the shell](https://hbctraining.github.io/Intro-to-Shell/lessons/01_the_filesystem.html) | 70 min |
|[Searching and redirection in shell](https://hbctraining.github.io/Intro-to-Shell/lessons/02_searching_files.html) | 45 min |
|[Introduction to the Vim text editor ](https://hbctraining.github.io/Intro-to-Shell/lessons/03_vim.html) | 30 min |
|[Shell scripts and `for` loops](https://hbctraining.github.io/Intro-to-Shell/lessons/04_loops_and_scripts.html) | 75 min |
|[Permissions and environment variables](https://hbctraining.github.io/Intro-to-Shell/lessons/05_permissions_and_environment_variables.html) | 50 min |
|[Introduction to ChIP-seq]() | 60 min |
|[Project setup and data organization](l) | 40 min |
|[Introduction to High-Performance Computing for HMS-RC's O2 cluster](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lectures/HPC_intro_O2.pdf) | 45 min |
|[Quality control and Alignment]() | 60 min |
|[Peak Calling with MACS2]() | 90 min |
|[ChIP quality assessment using cross-correlation]() | 90 min |
|[ChIP quality assessment using ChIPQC]() | 90 min |
|[Handling replicates]() | 75 min |
|[Functional analysis using ChipSeeker]() | 60 min |
|[Web-based fucntional analysis]() | 60 min |
|[Differential enrichment of peaks]() | 90 min |
|[Visualization of ChIP-seq data]() | 45 min |

### Dataset


***
*These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*

* *Some materials used in these lessons were derived from work that is Copyright Â© Data Carpentry (http://datacarpentry.org/). 
All Data Carpentry instructional material is made available under the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0).*

